content/blog/text-rendering-updates/index.Rmd

```diff
@@ -0,0 +1,171 @@
+
+But what if you don't want to think too much about all these details and just want to ensure that some specific font is available when a piece of code is running? In that case `require_font()` got you covered. This function allows you to state a dependency on a font in a script. The function scans the available fonts on the system and, if it doesn't find anything, proceeds to look for the font in the online repositories, downloading it if it finds it. If that also fails the function will either throw an error, or map the required font to a fallback of your choosing:
+
+```{r}
+#| fig.height: 3
+library(systemfonts)
+require_font("Rubik Moonrocks")
+
+plot.new()
+text(0.5, 0.5, "Fancy Font", family = "Rubik Moonrocks", cex = 6)
+```
+
+Remember that all of these niceties only goes into effect if you use a graphics device that uses systemfonts. For now that more or less means that you should use ragg (you should use ragg anyway so that is not a terrible requirement).
+
+## Getting to the Glyphs
+
+Most fonts these days are based on a vector outline. That means that they can be scaled smoothly to any size and doesn't take up a lot of storage space. It also means that there are polygons inside the font files and that these can be extracted! This is now possible with systemfonts and the new `glyph_outline()` and `glyph_raster()` functions.
+
+```{r}
+#| fig.width: 4.5
+#| fig.height: 4.5
+# Get the location of the glyph inside the font
+moonrocks <- font_info("Rubik Moonrocks")
+G <- glyph_info("G", path = moonrocks$path, index = moonrocks$index)
+
+# Extract the outline of the glyph and plot it
+outline <- glyph_outline(G$index, moonrocks$path, moonrocks$index, size = 400)
+grid::grid.path(
+  x = outline$x,
+  y = outline$y + 20, # To raise the baseline a bit
+  id = outline$contour,
+  default.units = "bigpts",
+  gp = grid::gpar(fill = "grey", col = "black", lwd = 4)
+)
+```
+
+Extracting them as polygons means that we can do all sorts of weird stuff with them if we so pleases:
+
+```{r}
+#| fig.width: 5.5
+#| fig.height: 4.5
+# Skew the glyph making it italic
+grid::grid.path(
+  x = outline$x + outline$y * 0.4,
+  y = outline$y + 20, # To raise the baseline a bit
+  id = outline$contour,
+  default.units = "bigpts",
+  gp = grid::gpar(fill = "grey", col = "black", lwd = 4)
+)
+```
+
+(real italic glyphs are designed to look good skewed, not just skewed versions of the regular glyphs)
+
+Remember how I said "most fonts" in the beginning of this section. There are still fonts that do not provide an outline, the prime example being most emoji fonts. The glyphs in such fonts are encoded as multiple bitmaps at fixed sizes (Microsofts emoji font going a different way by encoding them as SVGs). Since we can't get to the data as outlines we can instead extract it as a raster:
```

Comment:
Not all fonts use vector outlines. Emoji fonts, for instance, typically use bitmap images[^2]. Here's how to work with bitmap-based glyphs, extracting the data as a raster instead of a vector:

[^2]: Microsoft's emoji fonts are an exception, using SVG format instead of bitmaps.